compose.py

```diff
@@ -0,0 +1,305 @@
+    ('python3-pip'),
+    'pytorch':
+    ('python3-dev')
+}
+#### helper functions
+def log(msg, force=False):
+    if force or not FLAGS.quiet:
+        try:
+            print(msg, file=sys.stderr)
+        except Exception:
+            print('<failed to log>', file=sys.stderr)
+
+
+def log_verbose(msg):
+    if FLAGS.verbose:
+        log(msg, force=True)
+
+
+def fail(msg):
+    fail_if(True, msg)
+
+
+def fail_if(p, msg):
+    if p:
+        print('error: {}'.format(msg), file=sys.stderr)
+        sys.exit(1)
+
+
+##### create base image
+def make_dockerfile(ddir, dockerfile_name):
+    # Set enviroment variables
+    df = '''
+#
+# Multistage build.
+#
+FROM nvcr.io/nvidia/tritonserver:{}-py3 as full
+FROM nvcr.io/nvidia/tritonserver:{}-py3-min
+ARG TRITON_VERSION={}
+ARG TRITON_CONTAINER_VERSION={}
+ENV TRITON_SERVER_VERSION ${{TRITON_VERSION}}
+ENV NVIDIA_TRITON_SERVER_VERSION ${{TRITON_CONTAINER_VERSION}}
+ENV TRITON_SERVER_VERSION ${{TRITON_VERSION}}
+ENV NVIDIA_TRITON_SERVER_VERSION ${{TRITON_CONTAINER_VERSION}}
+LABEL com.nvidia.tritonserver.version="${{TRITON_SERVER_VERSION}}"
+ENV PATH /opt/tritonserver/bin:${{PATH}}
+'''.format(FLAGS.upstream_container_version, FLAGS.upstream_container_version, FLAGS.version, FLAGS.container_version)
+
+    # Copy over files
+    df += '''
+# Create a user that can be used to run triton as
+# non-root. Make sure that this user to given ID 1000. All server
+# artifacts copied below are assign to this user.
+ENV TRITON_SERVER_USER=triton-server
+RUN userdel tensorrt-server > /dev/null 2>&1 || true && \
+    if ! id -u $TRITON_SERVER_USER > /dev/null 2>&1 ; then \
+        useradd $TRITON_SERVER_USER; \
+    fi && \
+    [ `id -u $TRITON_SERVER_USER` -eq 1000 ] && \
+    [ `id -g $TRITON_SERVER_USER` -eq 1000 ]
+
+WORKDIR /opt/tritonserver
+RUN rm -fr /opt/tritonserver/*
+COPY --chown=1000:1000 --from=full /opt/tritonserver/LICENSE .
+COPY --chown=1000:1000 --from=full /opt/tritonserver/TRITON_VERSION .
+COPY --chown=1000:1000 --from=full /opt/tritonserver/NVIDIA_Deep_Learning_Container_License.pdf .
+COPY --chown=1000:1000 --from=full /opt/tritonserver/bin bin/
+COPY --chown=1000:1000 --from=full /opt/tritonserver/lib lib/
+COPY --chown=1000:1000 --from=full /opt/tritonserver/include/triton/core include/triton/core
+# Top-level include/core not copied so --chown does not set it correctly,
+# so explicit set on all of include
+RUN chown -R triton-server:triton-server include
+'''
+    with open(os.path.join(ddir, dockerfile_name), "w") as dfile:
+        dfile.write(df)
+
+
+### add additional backends needed
+def add_requested_backends(ddir, dockerfile_name, backends):
+    df = "# Copying over backends \n"
+    for backend in backends:
+        if (backend.lower()
+                in (EXAMPLE_BACKENDS + CORE_BACKENDS + NONCORE_BACKENDS)):
+            df += '''COPY --chown=1000:1000 --from=full /opt/tritonserver/backends/{} /opt/tritonserver/backends/{}    
+'''.format(backend, backend)
```

Comment:
Since everything is copied from the release image, I suggest currently this compose script doesn't support composition of custom builds? i.e. no GPU, no cloud storage, etc.?

Reply:
@GuanLuo can you enlighten me about what backends shouldn't be supported? We do need support for both CPU only and GPU only. Which backend is allowed for cloud storage?

Reply:
I think cloud storage is on the Triton core, and other backends mostly just have configuration on whether GPU should be supported. But the idea is that eventually we will need to have the Triton and other components be built in configuration as requested and compose with them instead of using the ones in the release container.

Reply:
Will change this for next round

Reply:
Not sure what is being asked here but the compose.py will do no builds. It will just take the entire triton core from the release image and include it.

Reply:
Is there something I need to configure for each backend so it won't run in GPU? Is @GuanLuo suggesting changing the backend builds so that they are optimized for CPU only?

Reply:
We can't work on the CPU-only support yet. In general we may need CPU-only builds of the backends (so a separate full image that is built for cpu only)